pkg/test/framework/components/zipkin/kube.go

```diff
@@ -0,0 +1,169 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zipkin
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sort"
+
+	"istio.io/istio/pkg/test/framework/components/environment/kube"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/resource"
+	testKube "istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/scopes"
+)
+
+const (
+	appName    = "zipkin"
+	tracesAPI  = "/api/v2/traces"
+	zipkinPort = 9411
+)
+
+var (
+	_ Instance  = &kubeComponent{}
+	_ io.Closer = &kubeComponent{}
+)
+
+type kubeComponent struct {
+	id        resource.ID
+	address   string
+	forwarder testKube.PortForwarder
+	env       *kube.Environment
+}
+
+func newKube(ctx resource.Context) (Instance, error) {
+	env := ctx.Environment().(*kube.Environment)
+	c := &kubeComponent{
+		env: env,
+	}
+	c.id = ctx.TrackResource(c)
+
+	// Find the zipkin pod and service, and start forwarding a local port.
+	cfg, err := istio.DefaultConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	fetchFn := env.Accessor.NewSinglePodFetch(cfg.SystemNamespace, fmt.Sprintf("app=%s", appName))
+	pods, err := env.Accessor.WaitUntilPodsAreReady(fetchFn)
+	if err != nil {
+		return nil, err
+	}
+	pod := pods[0]
+
+	forwarder, err := env.Accessor.NewPortForwarder(pod, 0, zipkinPort)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := forwarder.Start(); err != nil {
+		return nil, err
+	}
+	c.forwarder = forwarder
+	scopes.Framework.Debugf("initialized zipkin port forwarder: %v", forwarder.Address())
+
+	c.address = fmt.Sprintf("http://%s", forwarder.Address())
+	return c, nil
+}
+
+func (c *kubeComponent) ID() resource.ID {
+	return c.id
+}
+
+func (c *kubeComponent) QueryTraces() ([]Trace, error) {
+	// Get all traces
+	resp, err := http.Get(c.address + tracesAPI)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("zipkin api returns non-ok: %v", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	traces, err := extractTraces(body)
+	if err != nil {
+		return nil, err
+	}
+	return traces, nil
+}
+
+// Close implements io.Closer.
+func (c *kubeComponent) Close() error {
+	return c.forwarder.Close()
+}
+
+func extractTraces(resp []byte) ([]Trace, error) {
+	var traceObjs []interface{}
+	if err := json.Unmarshal(resp, &traceObjs); err != nil {
+		return []Trace{}, err
+	}
+	var ret []Trace
+	for _, t := range traceObjs {
+		spanObjs, ok := t.([]interface{})
+		if !ok && len(spanObjs) == 0 {
```

Comment:
should this be OR? "if it failed to parse or empty, skip"? print a debug-level warning about failing to parse the JSON.

Reply:
Right. Thanks for catching this!